### PR TITLE
Update to modern CMake (requires minimum 3.1.3 but recommends 3.3)

### PR DESCRIPTION
**Pull request by jrmadsen**
- Update to modern CMake (requires minimum 3.0.2 but recommends 3.3)
- Minor fixes to build dependencies in unit tests
- Plenty of comments
- Benefits:
  - "linking" to gotcha target in CMake provides `${gotcha_INCLUDE_DIRS}`
  - Improvements when GOTCHA is a submodule, particularly `set(CMAKE_C_EXTENSIONS ON)` which, in it's absence, will cause GOTCHA to fail to compile when a master project does not want `-std=gnuXX`
  - More modularity through INTERFACE libraries because they don't propagate their include-dirs, compile-definitions, etc. to everyone after
  - More portable method for checking for C++11 support

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
@jrmadsen : I like a lot of this
@daboehme: does it work with Caliper and apps which use Caliper?
@mplegendre: if you can maintain this code, it's good modern CMake, it shouldn't hurt codes that use the old style, opens up possibilities for apps that use modern CMake, and makes usage within an IDE better

I don't know what CMake we're/you're on, but I believe RAJA is in the high 3's now

**Comment by jrmadsen** [comment]
@DavidPoliakoff From the "external" project viewpoint, the only difference would be that linking to the `gotcha` CMake target would provide the include flags for the installation so if it were to break external projects, I would be very, very surprised.

**Comment by DavidPoliakoff** [comment]
@jrmadsen : yeah, I occasionally got a little careless in my development and burned the apps, so even if I know it can't hurt the apps I still like to test with them. I'd be astonished if the tests have problems, but it's still worth doing them

**Comment by jrmadsen** [comment]
Haha, yea you won't get any argument from me there

**Comment by jrmadsen** [comment]
Ok, Travis should be fixed now... Upgraded to 3.1.3 instead of 3.0.2 and export/installed the interface libraries from the top-level directory instead of src/. Looks like CMake is missing a policy notification about the latter issue when using newer versions of CMake bc it is supposed to "dumb" itself down and warn about things like that.

**Comment by jrmadsen** [comment]
Ok, now it's all good.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,50 +1,91 @@
...
+
+# set the compile flags
+target_compile_options(gotcha-compile-options INTERFACE "-fvisibility=${DEFAULT_SYMBOL_VISIBILITY}")
+# set the link flags
+set_target_properties(gotcha-compile-options PROPERTIES
+    INTERFACE_LINK_FLAGS "-fvisibility=${DEFAULT_SYMBOL_VISIBILITY}")
+
+# set gotcha-include to provide one path when building but another after installation
+target_include_directories(gotcha-include INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
```
Comment (DavidPoliakoff) [inline CMakeLists.txt]: This I didn't know about, that's really slick

**Comment by jrmadsen** [comment]
By the way, I have verified this works with an external package

**Pull request closed without merging**